--- NVIDIA-Linux-x86_64-304.125-no-compat32/kernel/nv.c~	2014-12-02 04:58:34.000000000 +0100
+++ NVIDIA-Linux-x86_64-304.125-no-compat32/kernel/nv.c	2015-03-30 21:52:46.978945686 +0200
@@ -2026,7 +2026,11 @@
     unsigned long i_arg
 )
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
+    return nv_kern_ioctl(file->f_path.dentry->d_inode, file, cmd, i_arg);
+#else
     return nv_kern_ioctl(file->f_dentry->d_inode, file, cmd, i_arg);
+#endif
 }
 
 long nv_kern_compat_ioctl(
@@ -2035,7 +2039,11 @@
     unsigned long i_arg
 )
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
+    return nv_kern_ioctl(file->f_path.dentry->d_inode, file, cmd, i_arg);
+#else
     return nv_kern_ioctl(file->f_dentry->d_inode, file, cmd, i_arg);
+#endif
 }
 
 /*
